Add EIP: EOF Contract Creation (#8209)
ethereum/EIPs: EIPS/eip-7620.md

---
eip: 7620

> title: EOF - Contract Creation Instructions

Reviewer: ```suggestion
title: EOF Contract Creation
```

since eip introduces both new instructions and a new tx type

> description: Introduce CREATE3, CREATE4, RETURNCONTRACT instructions and EOF contract creation logic

Reviewer: ```suggestion
description: Introduce `CREATE3`, `CREATE4`, `RETURNCONTRACT` instructions along with a new `InitcodeTransaction` transaction

author: Alex Beregszaszi (@axic), Paweł Bylica (@chfast), Andrei Maiboroda (@gumb0), Piotr Dobaczewski (@pdobacz)
discussions-to: https://ethereum-magicians.org/t/eip-7620-eof-contract-creation-instructions/18625
status: Draft
type: Standards Track
category: Core
created: 2024-02-12
requires: 170, 1559, 2028, 2718, 3540, 3541, 3670, 3860
---

## Abstract

> EVM Object Format (EOF) removes the possibility to create contracts using creation transactions (with an empty `to` field), and `CREATE` or `CREATE2` instructions. We introduce three new instructions: `CREATE3`, `CREATE4` and `RETURNCONTRACT`, as well as a new transaction type (`InitcodeTransaction`) to provide a way to create contracts using EOF containers.

Reviewer: ```suggestion
EVM Object Format (EOF) removes the possibility to create contracts using creation transactions (with an empty `to` field), `CREATE` or `CREATE2` instructions. We introduce three new instructions: `CREATE3`, `CREATE4` and `RETURNCONTRACT`, as well as a new transaction type (`InitcodeTransaction`) to provide a way to create contracts using EOF containers.
```

## Motivation
    
This EIP uses terminology from the [EIP-3540](./eip-3540.md) which introduces the EOF format.

EOF aims to remove code observability, which is a prerequisite to legacy EVM contract creation logic using create transactions, `CREATE` and `CREATE2`, because both the initcode and code are available to the EVM and can be manipulated. On the same premise, EOF removes opcodes like `CODECOPY` and `EXTCODECOPY`, introducing EOF subcontainers as a replacement to cater for factory contracts creating other contracts.
    
The new instructions and the new transaction type introduced in this EIP operate on EOF containers enabling all the use cases of contract creation that legacy EVM has.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
    
We introduce three new instructions on the same block number [EIP-3540](./eip-3540.md) is activated on:

1. `CREATE3` (`0xec`)
2. `CREATE4` (`0xed`)
3. `RETURNCONTRACT` (`0xee`)
    
### Transaction Types

> Introduce new transaction type `InitcodeTransaction` which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.

Reviewer: so the type still remains `2`?

> Introduce new transaction type `InitcodeTransaction` which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.

Reviewer: could you also list `TransactionPayload` fields here for better clarity

> Introduce new transaction type `InitcodeTransaction` which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.

Reviewer: ```suggestion
Introduce new transaction `InitcodeTransaction` (type `INITCODE_TX_TYPE`) which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.
```

The `initcodes` can only be accessed via the `CREATE4` instruction (see below), therefore `InitcodeTransactions` are intended to be sent to contracts including `CREATE4` in their execution.

We introduce a standardised Creator Contract (i.e. written in EVM, but existing at a known address, such as precompiles), which eliminates the need to have create transactions with empty `to`. Deployment of the Creator Contract will require an irregular state change at EOF activation block. Note that such introduction of the Creator Contract is needed, because only EOF contracts can create EOF contracts. See below for Creator Contract code.

Under transaction validation rules `initcodes` are not validated for conforming to the EOF specification. They are only validated when accessed via `CREATE4`. This avoids potential DoS attacks of the mempool. If during the execution of an `InitcodeTransaction` no `CREATE4` instruction is called, such transaction is still valid.

`initcodes` data is similar to calldata for two reasons:

1. It must be fully transmitted in the transaction.
2. It is accessible to the EVM, but it can't be fully loaded into EVM memory.

> For these reason `initcodes` data costs the same cost as for calldata (16 gas for non-zero bytes, 4 for zero bytes -- see [EIP-2028](./eip-2028.md)).

Reviewer: can we create a small section with a heading GAS SCHEDULE or somthing similar for better clarity and readability?

[EIP-3860](./eip-3860.md) and [EIP-170](./eip-170.md) still apply, i.e. `MAX_CODE_SIZE` as 24576, `MAX_INITCODE_SIZE` as `2 * MAX_CODE_SIZE`. Define `MAX_INITCODE_COUNT` as 256.

`InitcodeTransaction` is invalid if there are more than `MAX_INITCODE_COUNT` entries in `initcodes`, or if any one exceeds `MAX_INITCODE_SIZE`.

`InitcodeTransaction` is invalid if the `to` is `nil`.

#### RLP and signature

Given the definitions from [EIP-2718](./eip-2718.md) the `TransactionPayload` for an `InitcodeTransaction` is the RLP serialization of:

```
[chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, to, value, data, access_list, initcodes, y_parity, r, s]
```

`TransactionType` is `INITCODE_TX_TYPE` (`0x04`) and the signature values `y_parity`, `r`, and `s` are calculated by constructing a secp256k1 signature over the following digest:

```
keccak256(INITCODE_TX_TYPE || rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, to, value, data, access_list, initcodes]))
```

The [EIP-2718](./eip-2718.md) `ReceiptPayload` for this transaction is `rlp([status, cumulative_transaction_gas_used, logs_bloom, logs])`.

### Execution Semantics
    

> - The instructions `CREATE`, `CREATE2` are deprecated and rejected by validation in EOF contracts. They are only available in legacy contracts.

Reviewer: ```suggestion
- The instructions `CREATE`, `CREATE2` are made obsolete and rejected by validation in EOF contracts. They are only available in legacy contracts.
```

## Backwards Compatibility

This change poses no risk to backwards compatibility, as it is introduced at the same time EIP-3540 is. The new instructions are not introduced for legacy bytecode (code which is not EOF formatted), and the contract creation options do not change for legacy bytecode.
    
Legacy create transactions with `data` starting with `EF00` now are invalid. Similarly `CREATE` and `CREATE2` calls with `EF00` initcode fail early without executing the initcode. Previously, in both cases the initcode execution would begin and fail on the first undefined instruction `EF`.

## Test Cases

Creation transaction, `CREATE` and `CREATE2` cannot have its *code* starting with `0xEF`, but such cases are covered already in [EIP-3541](./eip-3541.md). However, new cases must be added where creation transaction, `CREATE` or `CREATE2` have its *initcode* being (validly or invalidly) EOF formatted as:
    

> - `0xEF` - initcode starts execution and fails

Reviewer: table format will be nice

    
The new instructions and the new transaction type introduced in this EIP operate on EOF containers enabling all use cases of contract creation that legacy EVM has.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
    
### Parameters

| Constant | Value |
| - | - |
| `INITCODE_TX_TYPE` | `Bytes1(0x04)` |
| `MAX_INITCODE_COUNT` | `256` |
| `GKECCAK256WORD` | Defined as `6` in the Ethereum Yellow Paper |
| `GCREATE` | Defined as `32000` in the Ethereum Yellow Paper |
| `GCODEDEPOSIT` | Defined as `200` in the Ethereum Yellow Paper |
| `GTXDATAZERO` | Defined as `4` in the Ethereum Yellow Paper |
| `GTXDATANONZERO` | Defined as `16` in the Ethereum Yellow Paper |
| `INITCODE_WORD_COST` | Defined as `2` in [EIP-3860](./eip-3860.md) |
| `MAX_INITCODE_SIZE` | Defined as `2 * MAX_CODE_SIZE` in [EIP-3860](./eip-3860.md) |

> | `MAX_CODE_SIZE` | Defined as `24576` in [EIP-170](./eip-170.md) |

Reviewer 2: As a suggestion, and not at all a hard requirement, you can link to the Yellow Paper or to EELS ([example](https://github.com/ethereum/execution-specs/blob/9ca885c8eb5e5151cda74f02e6675cd716764972/src/ethereum/shanghai/vm/interpreter.py#L61)) for these constants to provide more context.

Author: I have tried YP, that was rejected due to https://github.com/ethereum/EIPs/pull/8248. Didn't try EELS though, thanks!

3. `RETURNCONTRACT` (`0xee`)
    
### Transaction Types

Introduce new transaction `InitcodeTransaction` (type `INITCODE_TX_TYPE`) which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.

The `initcodes` can only be accessed via the `CREATE4` instruction (see below), therefore `InitcodeTransactions` are intended to be sent to contracts including `CREATE4` in their execution.

We introduce a standardised Creator Contract (i.e. written in EVM, but existing at a known address, such as precompiles), which eliminates the need to have create transactions with empty `to`. Deployment of the Creator Contract will require an irregular state change at EOF activation block. Note that such introduction of the Creator Contract is needed, because only EOF contracts can create EOF contracts. See below for Creator Contract code.

#### Gas schedule

Each `initcodes` item data costs the same as calldata (`GTXDATANONZERO` gas for non-zero bytes, `GTXDATAZERO` for zero bytes -- see [EIP-2028](./eip-2028.md)). The intrinsic gas of an `InitcodeTransaction` is extended by the sum of all those items' costs.

> 

Reviewer 2: Perhaps some pseudocode (or actual code based on EELS) would be nice here to make the gas costs less ambiguous?

    
| Initcode | Expected result |
| - | - |
| `0xEF` | initcode starts execution and fails |
| `0xEF01` | initcode starts execution and fails |
| `0xEF5f` | initcode starts execution and fails |
| `0xEF00` | invalid creation transaction (or `CREATE` / `CREATE2` fails early, returns 0 and keeps sender nonce intact) |
| `0xEF0001` | as above |
| valid EOFv1 container | as above |

Since EOF contract validation happens for all EOF containers during `CREATE4`, the following cases must be tested:
    
- `CREATE4` references a valid EOF `initcontainer` having only valid subcontainers
- `CREATE4` references legacy code, contract creation fails
- `CREATE4` references an invalid EOF `initcontainer`, contract creation fails
- `CREATE4` references a valid EOF `initcontainer` having an invalid EOF subcontainer somewhere in the subcontainer tree, contract creation fails
- `CREATE4` references a valid EOF `initcontainer` having legacy code somewhere in the subcontainer tree, contract creation fails
  
Cases for initcode calling "nested" `CREATE3` or `CREATE4` in various combinations
    
## Security Considerations

> Needs discussion.

Reviewer 2: ```suggestion
Needs discussion. <!-- TODO -->
```

If you use an HTML-style comment, the linter will complain when you move to review.
